### DISCUSSION TITLE
S1 RTC Local Resolution Weighting

### DISCUSSION THREAD

**Discussion by alexisahedo** [johntruckenbrodt/pyroSAR#234] [Q&A] [answered]
Dear @johntruckenbrodt,

I’ll like to seek your help in clearing some questions I have regarding LRW, as I’m seeking to make a national S1 mosaic.

1. 	Is it possible to use the resulting image of `pyroSAR.snap.util.geocode` to create a LRW composite?
2. 	Using this post as reference:  [step forum](https://forum.step.esa.int/t/spatial-reference-lost-after-applying-multitemporal-compositing/33090/12), I was hoping to use the module Create Stack of SNAP to make what I understand is a LRW composite image, but actually I’m not sure if I’m skipping other necessary steps, am I?
3. 	Does PyroSAR or S1_NRB have any implementations for massive LRW processing? If not, is there any interest in adding such tool? 
4. 	Do you have any recommendations for the implementation of LRW in a graph/python semi-automatization 

As always, I really appreciate all your support. 

Saludos.
Alexis A.

**Comment by johntruckenbrodt** [comment, marked as answer]
Dear @alexisahedo,
pyroSAR and S1_NRB do not directly offer an implementation but it is not too hard to achieve. With S1_NRB output it is pretty easy using STAC and xarray:
```python
import stackstac
import rioxarray
import numpy as np
from S1_NRB.metadata.stac import make_catalog

# create a STAC catalog from a directory with S1-NRB products
target = '/path/to/s1-nrb/products'
nrb_catalog = make_catalog(directory=target, silent=True)

# load the STAC items into an xarray object
ds = stackstac.stack(items=nrb_catalog,
                     dtype=np.dtype('float32'),
                     chunksize=(-1, 1, 1024, 1024),
                     xy_coords='center')

# get the linear-scaled gamma nought RTC backscatter
vv = ds.sel(band='vv-g-lin')

# get the local contributing area aka scattering area
lca = ds.sel(band='contributing-area')

# compute LRW and log-scale it
weights = 1 / lca
lrw = vv.weighted(weights=weights.fillna(0)).mean(dim='time')
lrw = 10 * np.log10(lrw)

# write the result to COG
lrw.rio.write_nodata(-9999, inplace=True)
lrw.rio.write_crs(lrw['proj:wkt2'].item(), inplace=True)
lrw.rio.to_raster(raster_path='/path/to/output.tif', driver='COG')
```
I am not sure whether the computation is 100% correct though, it is just quick shot I gave it some months ago.  
With pyroSAR you would have to check how to get the data into xarray. I haven't tried it myself yet. pyroSAR does not yet offer anything for STAC so loading a list of output files directly into an xarray object shoulf be easiest.

**Reply by alexisahedo** [reply]
Dear @johntruckenbrodt,

I appreciate your kind answer, I’ll try to make some a quick test and review the result. I’m using rasterio or Open Data Cube to make the raster into an xarray, so working with it should not be a big problem. 
Additionally, I tried using the SNAP’s “Create Stack” tool without luck, the error `[snap error nodeld "create stack" "has no geocoding"]` popped up sadly, I’m wondering if information has been lost in the format transformation from BEAM-DIMAP to GeoTIFF. I’ll continue trying to make it work somehow, using Radar data has proven to be a bigger challenge that I initially thought.

Saludos.

**Reply by johntruckenbrodt** [reply]
Dear @alexisahedo,
I am not familiar with SNAP's `Create Stack` but your idea about the file format could be the answer. The metadata in the BEAM-DIMAP files is quite comprehensive and not everything is written to GeoTIFF. It would really be interesting whether the two approaches return the same result. Please keep me posted if you end up comparing then. Don't give up on SAR just yet, it's great once you get the hang of it :wink:.
Cheers
